api/v1/logicalvolume_types.go

```diff
@@ -18,6 +18,22 @@ type LogicalVolumeSpec struct {
 	NodeName    string            `json:"nodeName"`
 	Size        resource.Quantity `json:"size"`
 	DeviceClass string            `json:"deviceClass,omitempty"`
+
+	// 'type' specifies the logical volume type.
+	// Can be 'thin-snapshot' or left empty in case of a regular volume.
+	// +kubebuilder:validation:Optional
+	Type string `json:"type"`
+
+	// 'source' specifies the logicalvolume name of the source; if present.
+	// +kubebuilder:validation:Optional
+	Source string `json:"source"`
+
+	//'accessType' specifies how the user intends to consume the snapshot logical volume.
+	// Allowed values are: "ro" (read-only) and "rw" (read-write).
+	// Since the kubernetes snapshots are Read-Only, set accessType as 'ro' to activate thin-snapshots as read-only volumes.
+	// On the other hand, set accessType as 'rw' to activate thin-snapshots as read-write volumes.
+	// +kubebuilder:validation:Optional
+	AccessType string `json:"accessType"`
```

Comment:
Since we are using `thin-snapshots` both as a PVC and VolumeSnapshots; we need a way to identify how we activate the volume differently for both the cases (Activation refers to setting permissions on how Logical Volume will be used; eg Read-Write or Read only).

`AccessType` specifies how the user intends to consume the snapshot logical volume.
Allowed values are: "ro" (read-only) and "rw" (read-write).
Since the kubernetes snapshots are Read-Only, set accessType as 'ro' to activate thin-snapshots as read-only volumes.
On the other hand, set accessType as 'rw' to activate thin-snapshots as read-write volumes.


On the other hand, the `Type` field mentions that the logical volume created is not a regular lv, but a `thin-snapshot` logical volume.
This will be useful for the future as we can have multiple types of logical volumes that we can create(For e.g `thick-snapshot`)

Reply:
Thanks for the suggestion! We can do that, but my only concern is the PVC-PVC clone scenario. 
A PVC clone and a PVC restore; both have `AccessType: rw` and `Type: thin-snapshot`.
In this case, creating a new Type: `thin-restore` will create confusion if we use it for the PVC Clone scenario.

Although, If we want to remove the `Accesstype` parameter, we might have to add a new `Type` field, named `thin-clone`.

So we can go ahead with 2 possible approaches as of now:

1. Keep both the parameters, i.e `Type` and `AccessType` (current approach)
2. Remove the `AccessType` parameter, and add two more `Type` fields; i.e we will have `thin-snapshot`, `thin-restore` and `thin-clone` types in LV CR.

What do you think?

Reply:
I prefer proposal 2 too.

Reply:
Thanks! I'll update accordingly :+1:

Reply:
I bet proposal2 too.

Reply:
One advantage of the existing approach (Type + Access Type) is that it tells the topolvm-node what needs to be done without it having to know how it will be used - a thin-snapshot needs to be created and the access type set. This way the topolvm-node does not need to know what the lvm logicalvolume is being used for, it just knows the properties and what it needs to do. The decisions are made at the time of creating the LogicalVolume by the controller, which IMO will abstract away the internals from the node plugin.
Restore Snapshot will internally create a thin-snapshot of the source snapshot so the type would be thin-snapshot again.
Volume Clone will internally create a thin-snapshot of the source volume so the type would be thin-snapshot again.

What do you think? Does the topolvm-node need to know the proposed usage of the LogicalVolume?

Reply:
@nbalacha 
Thank you for your post. Your post made me realize that the Type field looks unnecessary to tell LVMD "what it needs to do". At the moment, having a Source is synonymous with the need to create a snapshot. The AccessType field will remain. About the Type field, we can consider it again in the future when there is a demand to specify a copy way other than thin-snapshot. How is this?

Reply:
Thanks @llamerada-jp @nbalacha , In that case, should we go ahead with the current approach?

Reply:
I believe keeping the `Type` field is really helpful; as Linux LVM understands only 2 types of logical volumes

- Normal Volumes
- Snapshots

Keeping the `Type` field in the LogicalVolume CR, will give a good idea that this logical volume is a `snapshot` in the backend. 
Thoughts?

Reply:
My personal preference would be to keep both fields as it allows the LogicalVolume to be defined clearly. It would also make it easier to add the implementation for thick snapshots in the future.

Reply:
Isn't it enough to have `snapshot` boolean field (default is `false`)? Whether a volume is a thick volume or a thin volume can be distinguished by device class.

|volume type|what device class is for|original approach's `type` field|my approach's `snapshot` field|
|:--|:--|:--|:--|
|thick regular volume|thick volume|""|false|
|thick snapshot volume|thick volume|"thick-snapshot"|true|
|thin regular volume|thin volume|""|false|
|thin snapshot volume|thin volume|"thin-snapshot"|true|

Reply:
LVM allows taking a thick snapshot of a thin volume and vice versa, which is why that field was introduced. If that is not something we want to support, the above approach looks good to me.

Reply:
Hi @llamerada-jp , If that is the case, we can don't need the `snapshot` field too. 
If a LogicalVolume CR has a `Source` parameter, this means that it is a snapshot. And with accessType, we can determine whether is a VolumeSnapshot or a Restored-PVC/Cloned-PVC.

So, the new LogicalVolume CR:

```
Spec: topolvmv1.LogicalVolumeSpec{
			Name:        sname,
			NodeName:    node,
			DeviceClass: dc,
			Size:        snapSize,
			Source:      sourceVol,
			AccessType:  accessType,
		},
```

What do you think?

Reply:
Oops, I forgot this fact.

I'm ok to keep both fields. @llamerada-jp any thoght?

Reply:
So, we should stick with the current approach @satoru-takeuchi?

Reply:
@Yuggupta27 Sorry for confusion. I wrote my last comment before reading your second last comment.

Now I'd like to adopt your opinion [there](https://github.com/topolvm/topolvm/pull/463#discussion_r892763088).

Reply:
I understood that both `Type` and `AccessType` are needed for future expansion. I am aware that the allowed values are `thin-snapshot` or empty in this implementation. Thank you.

Reply:
I think removing the type field and keeping the accesstype is good. If everyone agrees, we can make the changes.

Reply:
Please go ahead. The Type field is not needed for now. it may need to be added when explicit control is requested, as in the post below.
https://github.com/topolvm/topolvm/pull/463/files#discussion_r891399102

Reply:
Thank you. Confirming that the final CRD will be as follows:

```
Spec: topolvmv1.LogicalVolumeSpec{
			Name:        sname,
			NodeName:    node,
			DeviceClass: dc,
			Size:        snapSize,
			Source:      sourceVol,
			AccessType:  accessType,
		},
```